### ISSUE TITLE
Windows: Can't get past the `Logging in and updating...` prompt

### ISSUE THREAD

**Issue by guihkx** [dmadisetti/steam-tui#74]
While `steam-tui` builds just fine on Windows nowadays, I can't get past this screen:

![image](https://github.com/dmadisetti/steam-tui/assets/626206/b62e65bd-7f1e-4bf8-b87c-c8ce6d574fc9)

It looks like `steam-tui` spawns `steamcmd.exe` just fine:

![image](https://github.com/dmadisetti/steam-tui/assets/626206/04535fc5-e19a-4587-87f0-9f7b31c27c6c)

But my guess is that when `steam-tui` sends the `login <user>` command to `steamcmd.exe`, that doesn't actually work... And yes, I have followed [what the README says](https://github.com/dmadisetti/steam-tui#usage).

I'm not a Windows programmer (though I occasionally use it for gaming), so I'm not sure how to debug this further...

_Originally posted by @guihkx in https://github.com/dmadisetti/steam-tui/issues/48#issuecomment-2128821866_

**Comment by guihkx** [comment]
> Now-days? Did it work before?

As far as I know, previously steam-tui wouldn't even build on Windows, right?

So what I meant to say is that, currently, steam-tui builds cleanly on Windows without any modifications.

>Try flushing your cache by deleting ~/.cache/steam-tui/*

That did nothing, unfortunately.

>I don't know if redirecting works on windows, but can you try steam-tui 2> error.log

Redirecting stderr to a file works (with PowerShell 7, at least), because I've added some `eprintln!()` calls in some places and they showed up in the file, but unfortunately none of the `log!()` messages you've added show up.

But I suppose I can replace every `log!()` call by `eprintln!()` and see what happens. I'll do that shortly.

**Comment by dmadisetti** [comment]
You can just update the log definition: https://github.com/dmadisetti/steam-tui/blob/f5d2a4f75ade23f21da18051a5469f2a084ec300/src/util/log.rs#L4

The conditional probably doesn't work on windows

**Comment by guihkx** [comment]
You mean like this, correct?

```diff
diff --git a/src/util/log.rs b/src/util/log.rs
index 6f2c41e..994d3c5 100644
--- a/src/util/log.rs
+++ b/src/util/log.rs
@@ -1,8 +1,6 @@
 macro_rules! log{
     ($($elem: expr),+ ) => {
-        if !atty::is(atty::Stream::Stderr) {
-            eprintln!(concat!($(concat!(stringify!($elem), " - {:?}\n")),+), $($elem),+);
-        }
+        eprintln!(concat!($(concat!(stringify!($elem), " - {:?}\n")),+), $($elem),+);
     };
 }
```

Unfortunately, `error.log` is empty...

**Comment by dmadisetti** [comment]
Yep!

**Comment by guihkx** [comment]
I've added this:

```diff
diff --git a/src/interface/steam_cmd.rs b/src/interface/steam_cmd.rs
index 2c7abad..d64d4c1 100644
--- a/src/interface/steam_cmd.rs
+++ b/src/interface/steam_cmd.rs
@@ -39,10 +39,12 @@ impl SteamCmd {
         let mut cmd = SteamCmd::with_args_and_seperator(args, 0x1b)?;

         // Send start up data I guess yeah?
+        eprintln!("calling next() 4 times...");
         cmd.next();
         cmd.next();
         cmd.next();
         cmd.next();
+        eprintln!("done calling next() 4 times.");

         Ok(cmd)
     }
```

And the log file only shows the first message, and never the second one, even after I waited a couple of minutes.

Do you think the issue is somehow related to that?

**Comment by dmadisetti** [comment]
Oh, it's empty. Are you using cargo run?

So the program is a little brittle. It's basically just writing and reading from steamcmd, and there's a chance it can hang if it doesn't think that it's received everything it should have from steamcmd.

The windows prompt response might be different in terms of line endings or special characters or something else. This is probably the most important log, since it should just show what the steamcmd output is:

https://github.com/dmadisetti/steam-tui/blob/f5d2a4f75ade23f21da18051a5469f2a084ec300/src/client.rs#L427

**Comment by guihkx** [comment]
>Are you using cargo run?

Nope, I'm building with `cargo build` and then running like `.\target\debug\steam-tui.exe 2> err.txt`


>and there's a chance it can hang if it doesn't think that it's received everything it should have from steamcmd.

Hmmm, I see.


>The windows prompt response might be different in terms of line endings or special characters or something else

That's actually a good guess! But I'll have look it up because I have no idea how the Windows console works.

>`log!(prompt);`

Yeah, that doesn't show up in the log file at all...

**Comment by dmadisetti** [comment]
Definitely a line scrubbing issue. Try putting a print between each. If the first one is failing, then remove the whole block?

You might be able to find the separator character in windows. The 0x1b is what's used in linux and is the escape character: http://www.csc.villanova.edu/~tway/resources/ascii-table.html

https://github.com/dmadisetti/steam-tui/blob/f5d2a4f75ade23f21da18051a5469f2a084ec300/src/interface/steam_cmd.rs#L39

**Comment by dmadisetti** [comment]
If you can invoke steamcmd, play around with it (examine some game IDs) and send me the raw file, I might be able to help you further. I'm just guessing as is

does power shell have `tee` https://stackoverflow.com/questions/52970939/right-usage-of-21-tee-or-tee-with-powershell?

**Comment by guihkx** [comment]
Thank you for all the suggestions!

After investigating and learning a bit more about Windows, I' afraid this will not be feasible with only `std::process::Command` + `std::io:BufReader`.

I might be off on the actual technical details, but from what I understand, the issue was never about lack of support for ANSI escape characters, as those were added in Windows 10. Instead, the issue is related to how processes are spawned and handled on Windows compared to Linux.

Windows 10 [introduced ConPTY](https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/), which I think is aimed to behave closer to what we have on Linux.

This is just my guess, but this won't work without ConPTY because whatever Windows uses to spawn processes by default, causes stdout to be buffered until a new line character gets outputted, and so that's why we never reach the `Steam>` prompt on Windows.

I've been experimenting with [this dependency](https://docs.rs/crate/conpty/latest), and I wrote a test program that spawns `steamcmd` on this ConPTY thing, and at least now the `Steam>` prompt can be seen, and therefore we can jump to the `0x1b` separator character as well.

Anyway, I'm still not 100% sure if this will work, but it does looks very promising.

The next step, I suppose, is implementing `conpty` on `steam-tui` itself. I can give it a shot, but I'm also not a very good Rust programmer, so it's very likely it will be a crap implementation. :)

**Comment by dmadisetti** [comment]
Hmmm, very interesting.

I say go for it, learning rust is fun and a good skill to have.

~I do have a suggestion that's a bit more of a hack:
Input a new line before every read kind of like this block that just keeps writing "" until it gets the it detects the prompt. Maybe experiment with writing new lines too.~

Nvm, looks like conpty is part of this https://github.com/zhiburt/expectrl?tab=readme-ov-file
I think this could be a better solution to the handrolled code I have right now.

Happy to review any conpty solutions - but this might be a quick work around.

**Comment by Megaguy32** [comment]
I found logging in with `steamcmd +login <username>` helps prod steam-tui along.
In case it's relevant, I was prompted to input the 2fa code from the steam guard app.
